### ISSUE TITLE
ELK behind DFP

### ISSUE THREAD

**Issue by mborejdo** [vfarcic/docker-flow-proxy#225]
I am running an ELK stack in my swarm. I want the service to be available on two serviceDomains, proxying to different ports.
For example, I want kibana (port 5601) to be available on kibana.hostname and elasticsearch (port 9200) on elasticsearch.hostname.
I have tried different combinations of srcPort, port, etc, with no success.

My current stack config looks like this:

```
[...]
      labels:
        - com.df.notify=true
        - com.df.serviceDomain.1=kibana.hostname
        - com.df.port.1=5601
        - com.df.srcPort.1=80
        - com.df.serviceDomain.2=kibana.hostname
        - com.df.port.2=5601
        - com.df.srcPort.2=443
        - com.df.serviceDomain.3=elasticsearch.hostname
        - com.df.port.3=9200
        - com.df.srcPort.3=80
        - com.df.port.4=9200
        - com.df.srcPort.4=443
        - com.df.distribute=true
```

**Comment by leostarcevic** [comment]
Multiple domains should be separated with comma (,)
http://proxy.dockerflow.com/usage/

**Comment by mborejdo** [comment]
@leostarcevic I am exposing port 5601 and 9200 from one service. as far as I understand you need to use the "postfix"-notation (the docs seem to be here a bit unclear).
Using multiple serviceDomains for one service (and one destination port) works with a commaseperated list, though

**Comment by leostarcevic** [comment]
Ok, I'm not sure I understood the issue here, did your service labels got registered by dfp?
What does the /cfg/haproxy.cfg end up like inside the proxy service?

**Comment by leostarcevic** [comment]
You are right about the "postfix" - notation.. re-reading the docs:

Multiple destinations for a single service can be specified by adding index as a suffix to serviceDomain, srcPort, and port parameters. In that case, srcPort is required. Defining multiple destinations is useful in cases when a service exposes multiple ports with different paths and functions.

**Comment by mborejdo** [comment]
@leostarcevic currently I cannot share my full compose-yaml, since its very specific. I will try to prepare a stripped-down example.

Using these labels

```
      labels:
        - com.df.notify=true
        - com.df.serviceDomain=kibana.hostname
        - com.df.port=5601
        - com.df.srcPort=443
        - com.df.serviceDomain.2=elasticsearch.hostname
        - com.df.port.2=9200
        - com.df.srcPort.2=80
        - com.df.distribute=true
```

generates a haproxy.cfg like this
```
[...]
frontend services
    acl url_elastic_logstash5601 path_beg /
    acl srcPort_elastic_logstash443 dst_port 443
    acl domain_elastic_logstash hdr(host) -i kibana.hostname
    use_backend elastic_logstash-be5601 if url_elastic_logstash5601 domain_elastic_logstash srcPort_elastic_logstash443

[...]

backend elastic_logstash-be5601
    mode http
    log global
    server elastic_logstash elastic_logstash:5601

```

Suffixed dfp-labels seem to be ignored

**Comment by vfarcic** [comment]
@mborejdo Can you confirm that you're using service (not container) labels? Are they inside `deploy` section?

**Comment by mborejdo** [comment]
My full yml, redacted:

```
version: "3"

services:

  logstash:
    image: java8-logstash
    hostname: logstash
    healthcheck:
      test: "curl --fail http://127.0.0.1:9200 || exit 1"
      interval: 2m
      timeout: 60s
      retries: 5
    networks:
        - EM
    ports:
        - 12201:12201
        - 25826:25826
    environment:
        ES_JAVA_OPTS: "-Xms4G -Xmx4G"
    volumes:
      [...]
    deploy:
      placement:
        constraints:
          - node.labels.de.XXX.stability == testing
          - node.hostname == em-sv-dnodealpha
      labels:
        - com.df.notify=true
        - com.df.serviceDomain=kibana.XXX.de
        - com.df.port=5601
        - com.df.srcPort=443
        - com.df.serviceDomain.2=elasticsearch.XXX.de
        - com.df.port.2=9200
        - com.df.srcPort.2=80
        - com.df.distribute=true

networks:
  EM:
    external: true


```

**Comment by leostarcevic** [comment]
Like I mentioned before, Iseems like you put everything under one of the services?

serviceDomain for elasticsearch and kibana are under the logstash service?

**Comment by mborejdo** [comment]
@leostarcevic correkt. This is one service, running as "ELK"-Stack. (elastic, logstash, kibana).
I know, its not the one-service-per-container-approach, but it works well for me. So yea, everything is one service with multiple ports (and endpoints)

**Comment by vfarcic** [comment]
Judging by the proxy config you posted, it's forwarding all requests (no matter the path) that are entering through port `443` and domain `kibana.hostname` and forwarding them to the service `elastic_logstash` on port `5601`. Can I assume that:

1. You have valid certs and that they are included in the proxy config?
2. You confirmed that elastic_logstash is reachable on port `5601` from inside a container in the same network?

**Comment by leostarcevic** [comment]
I'm not sure if it matters but shouldn't:
"com.df.serviceDomain" be "com.df.serviceDomain.1" ?

Also I'd recommend you to split your stack into three services, Viktor has a great example for the ELK-stack
https://github.com/vfarcic/docker-flow-stacks/blob/master/logging/logging.yml

**Comment by mborejdo** [comment]
@vfarcic I can confirm:

1. I have valid LE certs for kibana.redacted.de
I can reach kibana under 
https://kibana.redacted.de

I cannot reach kibana under 
http://kibana.redacted.de

This is expected, as far as i understand.

What i want to achieve is to reach
elasticsearch under

http://elasticsearch.redacted.de

which sould forward the request to logstash:9000 (inside the swarm).

This does not seem to work.

@leostarcevic 
I tried with .1 postfixed config, but that did not work at all. In that case also https://kibana.redacted.de was not available. So currently I assume there seems to be a problem parsing postfixed labels.

**Comment by vfarcic** [comment]
@mborejdo You're missing `.1` in your labels. DFP will start looking for labels with suffix `.1`. If it exists it'll look for suffix `.2`, and so on. Can you change your labels that do not have a suffix? For example, `com.df.port` should be `com.df.port.1`.

**Comment by mborejdo** [comment]
@vfarcic using this label config:

```
      labels:
        - com.df.notify=true
        - com.df.serviceDomain.1=kibana.redacted.de
        - com.df.port.1=5601
        - com.df.srcPort.1=443
        - com.df.serviceDomain.2=elasticsearch.redacted.de
        - com.df.port.2=9200
        - com.df.srcPort.2=80
        - com.df.distribute=true
```

generates no elastic related entries in haproxy.cfg.

removing the elk-stack reloads the config. readding the stack with adjusted labels, seems to do nothing.
Atleast the services are not available thorugh dfp (since no entries were generated). As stated, the un-postfixed entries worked before the restart.

**Comment by vfarcic** [comment]
@mborejdo Would it be possible to strip the image from any propietary parts and push it to Docker Hub? That would help me debug the problem and find the solution much faster.

**Comment by mborejdo** [comment]
@vfarcic the image is pretty basic. Its basically this image https://hub.docker.com/r/sebp/elk/ with some installed plugins. So nothing special here. Also my compose-yaml was pretty much the one I use (besides some removed volumes) So there is really nothing special here, as far as I can see

**Comment by vfarcic** [comment]
I created a simple stack using that image and got *Unable to connect to Elasticsearch at http://localhost:9200* error message. You can see the stack and the commands I executed in https://github.com/vfarcic/docker-flow-stacks/blob/master/logging/xxx.yml . This stack is still not using DFP (first I wanted to make sure it works in it's simplest form). Since you're familiar with that image you probably know what the problem is. Can you please let me know what I'm doing wrong?

**Comment by mborejdo** [comment]
@vfarcic thanks for your effort. Did you see any startup-logging for the container? The most-common problem seems to be this one: http://elk-docker.readthedocs.io/#es-not-starting-max-map-count

**Comment by vfarcic** [comment]
Does the stack I posted works as is on your system? If it doesn't, can you update it (definition of commands in comments). If we make the stack work without a proxy, I'm sure that it should not be a problem to hook it into DFP.

**Comment by mborejdo** [comment]
@vfarcic thanks for having a look. I will try the stack tomorrow, since i currently do not have access to my swarm. But my question is rather not ELK specific. I can confirm my stack it works, as it is in use for some months now. Before I used nginx as a swarm-entering proxy.
Could you maybe provide a "DFP multi-service" example (with suffixed labels) stack? I think any application offering multiple ports (other than 80 and 443) would work as an example.

**Comment by vfarcic** [comment]
Unfortunately, I don't have a container with multiple services. While there are probably good use-cases for such a decision, it is, more often than not, an anti-pattern. All the services I own are single service per container. In some cases there are multiple processes but only one of them has a public facing API.

Even though I cannot show you on one of my services, the logic you used is correct except for `serviceDomain`. At the moment, only `servicePath`, `serviceDomain`, `srcPort`, and `port` parameters can be indexed (e.g. `com.df.port.1`, `com.df.port.2`).

From the example you sent...

```
      labels:
        - com.df.notify=true
        - com.df.serviceDomain=kibana.hostname
        - com.df.port=5601
        - com.df.srcPort=443
        - com.df.serviceDomain.2=elasticsearch.hostname
        - com.df.port.2=9200
        - com.df.srcPort.2=80
        - com.df.distribute=true
```

... since there cannot be multiple domains, both elasticsearch and kibana would be accessible through the domain `kibana.hostname`. Can you please confirm that's not the case?

Ofcourse, the fact that (excluding `serviceDomain`) the labels look correct does not mean that they really are. That's why I'd like to try it myself but for that I need to create a service myself. If, when I run it on my laptop it doesn't work, I can debug it easily and get back to you with the solution.

As for using multiple service domains, DFP can easily be extended to do that. However, since it is not a common case (not many would use it), I would like to confirm that what you're doing works with a single domain.

**Comment by mborejdo** [comment]
@vfarcic I am not sure I understand your last comment. I am not sure how I would access kibana and elasticsearch thorugh the "kibana.hostname".  I am not exposing ports in this case. I would expect both services to be available on port 80 on different serviceDomains.
I can confirm that kibana is accessible on https://kibana.hostname and http://kibana.hostname. http://elasticsearch.hostname and https://elasticsearch.hostname show DFP errorpages. Using a label-config with only postfixed labels (serviceDomain, srcPort and port) fails registering with DFP.

Considering the stack you posted, it successfully started for me

I needed to run ` sudo sysctl -w vm.max_map_count=262144` first, though. Also the first init didnt work, so after the failed service got restarted it worked. So your example stack works as some multi-service example.

**Comment by vfarcic** [comment]
My comment on accessing both services through the same domain was based on (probably unintentional) usage of your labels.

```
      labels:
        - com.df.notify=true
        - com.df.serviceDomain=kibana.hostname
        - com.df.port=5601
        - com.df.srcPort=443
        - com.df.serviceDomain.2=elasticsearch.hostname
        - com.df.port.2=9200
        - com.df.srcPort.2=80
        - com.df.distribute=true
```

Such a combination means that you want to access:
1. Service through domain `kibana.hostname`, through port `443`, and have requests forwarded to `5601`.
2. Service through domain `elasticsearch.hostname`, through port `80`, and have requests forwarded to `9200`.

Is that what you're trying to do? Access one only through port `443` and the other only through port `80`?

If, as you said before, you'd like to access both services on port 80, `com.df.srcPort=443` label should be `com.df.srcPort=80` but that would fail without assigning different domains for each of them.

Also, `com.df.serviceDomain.2` (and other labels that end with `.2`) should be changed to `.1` (it must start with `1`).

So, the problem (apart from corrections to your stack file) is that service domains (there can be multiple) are applied to the whole service (in Docker terms), not divided by different processes inside a service (by different processes I mean ES, Kibana, and so on).

I need to think whether it would be good for the project to add the option to assign multiple domains tied to different processes since it would add complexity (however small) and would not be used by many (if anyone). I've been teaching and preaching that what you're trying to do is a big anti-pattern. I would strongly suggest that you split those processes into separate services. Otherwise, you're fighting against the logic behind containers and, in this case, even against how Elastic thinks things should be deployed.

**Comment by mborejdo** [comment]
>Such a combination means that you want to access:
>1. Service through domain kibana.hostname, through port 443, and have requests forwarded to 5601.
>2. Service through domain elasticsearch.hostname, through port 80, and have requests forwarded to 9200.

Thats exactly what I want to achieve (maybe even including a rewrite from http to https for the kibana service)

I understand that running multiple services withing one container is an antipattern. I am not sure if this is really the case here. I am not a ES expert, but as far as I understand the ELK stack elasticsearch is the base engine (exposing some http JSON API on 9200), which is used also by kibana (some plugin on top of ES), which provides an UI on port 5601 (I see this as some admin UI to ES), additionally there is logstash, some kind of logforwarder, which also exposes ports and uses ES. (these i do not want to expose)

As the used image also provides a way to start only a single service, I might think about that, but regarding the above case, I think using a "Bundled Elastic stack" it is not that "exotic"

**Comment by vfarcic** [comment]
ES is, indeed, the base of almost all Elastic products. Configuring ES in DFP is not a problem. Even multiple domains for a single service is not a problem. The issue is to have multiple domains tied to different processes inside the same service.

Bundled elastic stack is very common (e.g. https://github.com/vfarcic/docker-flow-stacks/blob/master/logging/logging-df-proxy.yml). What is not common is to have a single service/container with all those processes. Actually, I don't know whether it's common or not (it might be just as JBoss is still common). What I do know is that it shouldn't be. The reason is very simple. It does not scale. ES is a very resource hungry engine. Sooner or later you'll need to scale it. With the solution you're using that is not possible. You'd need to scale everything. On top of that, it's slow, Swarm has a hard time scheduling it, it is not fault-tolerant, and so on and so forth. On top of all that, having everything in the same container does not have any advantage (only disadvantages). You can easily accomplish the same with a stack file (e.g. https://github.com/vfarcic/docker-flow-stacks/blob/master/logging/logging-df-proxy.yml). If you do, Swarm will schedule services much better (especially if you put constraints) and do all the magic Docker does for us.

I hope you did not misunderstand my comments. I'm sure that you have good reasons for using such an image. I'm just trying to explain (in a very indirect way) the logic behind DFP. It is a tool meant to work with a very specific combination of principles. It is not meant to solve all the possible combinations people might have. One of the main motivations behind the project is to provide easy proxy configuration for a highly dynamic cluster where services are being created, updated, and removed in a relatively high frequency.

**Comment by mborejdo** [comment]
@vfarcic I understand your points and they really make me rethink my current usage of the ELK stack, Still I am curious what is going wrong in my case, as I cannot make postfixed labels work at all.
Lets put ELK aside and say I exposed my HTTP-Server on multiple ports and I would want to route different serviceDomains to different ports of my webserver-Container.

**Comment by vfarcic** [comment]
@mborejdo The problem is with `serviceDomain`. It applies to the whole service and is not tied to a specific port. An example would be:

```
      labels:
        - com.df.notify=true
        - com.df.serviceDomain=acme.com
        - com.df.port.1=1234
        - com.df.srcPort.1=80,443
        - com.df.reqMode.1=http
        - com.df.port.2=4321
        - com.df.srcPort.2=9200
        - com.df.reqMode.2=tcp
        - com.df.distribute=true
```

This example would reconfigure the proxy to forward all requests coming from `acme.com` and port `80` or `443` to the service these labels belong to and the port `1234`. It would, also, forward requests coming from `acme.com` and port `9200` to the port `4321`.

This was a mockup of a relatively common use case (an HTTP service that also accepts TCP requests). The important part is that `acme.com` is used for all ports. It is not limited to a particular port.

There can be multiple `serviceDomain` entries (e.g. `com.df.serviceDomain=acme1.com,acme2.com`) but they are still applied to the whole service, not a particular port.

In other words, DFP does not support `to route different serviceDomains to different ports`. It applies domain to the whole service no matter how many ports it uses.

This can be changed. I can extend DFP to tie a domain to a port. However, I cannot tell you when would that happen. There are other issues I need to fix and features to add. Since this scenario is very uncommon it would get low priority.

Would you be open to a PR. I would love if you could take over this feature and submit a PR. I could guide you through the code and the tests.

**Comment by mborejdo** [comment]
@vfarcic thank you very much for your explanation. I understand the problem now and most likely change my stack-config anyways, I will close this issue.

Considering the  PR and a possible guidance: if you think this would be a good start for a go newbie (exactly 0 LOC), I am all in (I am a mostly JS developer with some interest in devops). I will fork and have a look (trying to get tooling working first), then I will contact you with questions :-) 
As I work fulltime, this would be rather in the evening or next weekend 

Again thank you very much for your help and explanations.

**Comment by vfarcic** [comment]
DFP  might be a bit complicated. If you'd like to try GoLang, I'd suggest https://github.com/vfarcic/docker-flow-swarm-listener. It is a simple code that should be easy to understand. In any case, if you choose any of those I'd be more than happy to help you out.
